### PR TITLE
perf!: Optimize Filter Initialization with Concurrent Processing

### PR DESCRIPTION
**Pull request by qfrank**
## Summary
This PR includes several improvements to the messenger's filter initialization and chat preview functionality:

Key Changes:
- Refactored `InitFilters` into smaller, more focused functions for better maintainability, and init filters concurrently
- Moved filter initialization logic from `messenger.go` to a new file `messenger_filter_init.go`
- Added optional community chat filtering to `ChatsPreview` method
- Made history archives import asynchronous via `startHistoryArchivesImportLoop`

The main improvements are:
1. Better code organization by splitting the large `InitFilters` function into smaller, focused functions
2. Enhanced error handling and logging throughout the filter initialization process
3. Added ability to filter chat previews by community vs non-community chats
4. Improved performance by making history archives import non-blocking

The changes maintain backward compatibility while providing new filtering capabilities for chat previews.

With this PR, the duration for api `StartMessenger` has been reduced from 1.71s to 1.02s running with iphone XS Max. Relate mobile [issue](https://github.com/status-im/status-mobile/issues/21456)

## Why added ability to filter chat previews by community vs non-community chats?
Let's take a look at `chatsPreviewResponse.log`, the response is very long which contains around 3.77 million characters. There's lots of chats related to community, when mobile frontend loads the chat preview list, actually it doesn't need the community chats for the chat list data soon, so we can split the long response into 2 parts(one with community chats one with non-community chats). With this solution, we can see the chat preview list almost in real time. It need wait around 2-3 seconds before this solution. But if we want to fix the performance issue totally, I'm wondering if we can use `protobuf` as data transfer mechanism rather than `json` 🤔, you can see there's lots of string starts with `0x...` like `0x04ffd68fe2bc7b2fceff949fd06c0f7a181ee5df2fc7ce9960c73a08e3c316a196dbae1700855a038bafe5091ab2cee9ac6fa77ed9bffe2053deeb97083d75d59f` @igor-sirotin @ilmotta 
[chatsPreviewResponse.log.zip](https://github.com/user-attachments/files/17796589/chatsPreviewResponse.log.zip)

### CODE REVIEW

File: protocol/messenger_filter_init.go [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+
+	// Start all goroutines
+	wg.Add(5)
+	go m.processJoinedCommunities(&wg, filtersCh, errCh)
+	go m.processSpectatedCommunities(&wg, filtersCh, errCh)
+	go m.processChats(&wg, filtersCh, publicKeysCh, errCh)
+	go m.processContacts(&wg, publicKeysCh, errCh)
+	go m.processControlledCommunities(&wg, errCh)
+
+	// Wait for all goroutines to complete and close channels
```
Comment (ilmotta) [inline protocol/messenger_filter_init.go]: This type of comment is unnecessary because I'd imagine every Go developer knows what `Wait` and `close` do.

File: protocol/messenger_filter_init.go [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+	wg.Wait()
+	close(filtersCh)
+	close(publicKeysCh)
+
+	// Check for errors
```
Comment (ilmotta) [inline protocol/messenger_filter_init.go]: This comment seems superfluous.

File: protocol/messenger.go [hunk 3]
```diff
@@ -920,20 +917,7 @@ func (m *Messenger) Start() (*MessengerResponse, error) {
...
-	for _, joinedCommunity := range joinedCommunities {
-		// resume importing message history archives in case
-		// imports have been interrupted previously
-		err := m.resumeHistoryArchivesImport(joinedCommunity.ID())
-		if err != nil {
-			return nil, err
-		}
-	}
-	m.enableHistoryArchivesImportAfterDelay()
+	m.startHistoryArchivesImportLoop()
```
Comment (igor-sirotin) [inline protocol/messenger.go]: Keep the function simple, no need to nest the `go ...` inside
```go
func (m *Messenger) historyArchivesImportLoop() {
    defer gocommon.LogOnPanic()
    joinedCommunities, err := m.communitiesManager.Joined()
    // ...
}
```

And then start as just:

```suggestion
	go m.historyArchivesImportLoop()
```

File: protocol/messenger.go [hunk 4]
```diff
@@ -974,6 +958,28 @@ func (m *Messenger) Start() (*MessengerResponse, error) {
...
+			return
+		}
+
+		for _, joinedCommunity := range joinedCommunities {
+			// resume importing message history archives in case
+			// imports have been interrupted previously
+			err := m.resumeHistoryArchivesImport(joinedCommunity.ID())
+			if err != nil {
+				m.logger.Error("failed to resume history archives import", zap.Error(err))
+				return
```
Comment (igor-sirotin) [inline protocol/messenger.go]: I guess it makes sense to `continue` in this case?

File: protocol/messenger.go [hunk 5]
```diff
@@ -1775,221 +1781,6 @@ func (m *Messenger) handlePushNotificationClientRegistrations(c chan struct{}) {
 	}()
 }
 
-// InitFilters analyzes chats and contacts in order to setup filters
-// which are responsible for retrieving messages.
-func (m *Messenger) InitFilters() error {
```
Comment (igor-sirotin) [inline protocol/messenger.go:1739]: i like your intention to move the code to a separate file, but it makes it impossible to review. I suppose you didn't rewrite 200 lines, but diff view makes it look like you did.

Can you please move it back? 😄  And then move it to a separate file with a separate PR.

Comment (ilmotta) [reply protocol/messenger.go:1739]: I agree it would facilitate review to follow your suggestion @igor-sirotin, but would simply rewriting the history with separate commits and keep one PR be enough since this PR is relatively simple if we consider a good chunk of the code is identical to develop?

Usually my personal approach is to have a separate commit for file moves and renames in an attempt to let git understand what's going on. Then separate commits for each refactor technique applied.

File: protocol/messenger_chats.go [hunk 6]
```diff
@@ -46,10 +46,19 @@ func (m *Messenger) Chats() []*Chat {
 	return chats
 }
 
-func (m *Messenger) ChatsPreview() []*ChatPreview {
+// ChatsPreview returns a list of chat previews.
```
Comment (ilmotta) [inline protocol/messenger_chats.go]: It would be easier to understand and less prone to bugs if we didn't use a tri-state boolean. While integrating with this PR https://github.com/status-im/status-mobile/compare/ilmotta/use-optimized-filter-initialization?expand=1 I had to come back to this comment multiple times to make sure I wasn't misunderstanding anything. A better solution here I think is to use an enum, which would be self-explanatory and wouldn't require a comment.

In the client layer I had to create 3 constants to describe what the tri-state flag means:

```clojure
(def ^:const chat-preview-type-any nil)
(def ^:const chat-preview-type-community true)
(def ^:const chat-preview-type-non-community false)
```

File: services/ext/api.go [hunk 7]
```diff
@@ -291,8 +291,8 @@ func (api *PublicAPI) Chats(parent context.Context) []*protocol.Chat {
 	return api.service.messenger.Chats()
 }
 
-func (api *PublicAPI) ChatsPreview(parent context.Context) []*protocol.ChatPreview {
-	return api.service.messenger.ChatsPreview()
+func (api *PublicAPI) ChatsPreview(parent context.Context, onlyCommunityChats *bool) []*protocol.ChatPreview {
```
Comment (ilmotta) [inline services/ext/api.go]: I had the impression the standard was to use structs for the request args in endpoints because then we can always add keys (and thus behavior) without breaking clients. Is there an agreement on this convention? cc @igor-sirotin

File: protocol/messenger.go (hunk 5 above)
Comment (igor-sirotin) [reply protocol/messenger.go:1739]: Yes, separate commit with code movement works as well 👍

File: services/ext/api.go (hunk 7 above)
Comment (igor-sirotin) [reply services/ext/api.go]: @ilmotta this is true, but only for `mobile/status.go`. 

APIs in `services/**/api.go` are called through JSON-RPC server, so the behaviour in this case is a bit different, as we pass arguments as a JSON array. 

# TLDR;

Yes, for compatibility it is better to make `struct` arguments.

# Investigated details

I checked the behaviour for redundant/missing arguments.

1. **Redundant argument** - returns proper error

    Request: 
    ```json
    {
        "jsonrpc": "2.0",
        "method": "settings_getSettings",
        "params": ["redundant_arg"]
    }
    ```
    Response:
    ```json
    {
        "jsonrpc": "2.0",
        "id": 0,
        "error": {
            "code": -32602,
            "message": "too many arguments, want at most 0"
        }
    }
    ```

2. **Missing pointer argument** - uses `nil` value for missing argument

    Request:
    ```json
    {
        "jsonrpc": "2.0",
        "method": "wakuext_acceptContactRequest",
        "params": []
    }
    ```
    
    Response:
    ```json
    {
        "jsonrpc": "2.0",
        "id": 0,
        "error": {
            "code": -32603,
            "message": "method handler crashed"
        }
    }
    ```
    
    There was panic, because `request` argument is `nil` in
    https://github.com/status-im/status-go/blob/17dbc379641f9c54772df09ab10a4979ad9e968d/services/ext/api.go#L857-L859

3. **Missing non-pointer argument** - returns proper error
    
    `SetUserStatus` expects 2 arguments:
    https://github.com/status-im/status-go/blob/17dbc379641f9c54772df09ab10a4979ad9e968d/services/ext/api.go#L802-L804
    
    Request wihout `customText`:
    ```json
    {
        "jsonrpc": "2.0",
        "method": "wakuext_setUserStatus",
        "params": [
            1
        ]
    }
    ```
    
    Response:
    ```json
    {
        "jsonrpc": "2.0",
        "id": 0,
        "error": {
            "code": -32602,
            "message": "missing value for required argument 1"
        }
    }
    ```

File: protocol/messenger.go (hunk 3 above)
Comment (qfrank) [reply protocol/messenger.go]: fixed

File: protocol/messenger.go (hunk 4 above)
Comment (qfrank) [reply protocol/messenger.go]: fixed

File: protocol/messenger_filter_init.go (hunk 1 above)
Comment (qfrank) [reply protocol/messenger_filter_init.go]: fixed

File: protocol/messenger_filter_init.go (hunk 2 above)
Comment (qfrank) [reply protocol/messenger_filter_init.go]: fixed

File: protocol/messenger_chats.go (hunk 6 above)
Comment (qfrank) [reply protocol/messenger_chats.go]: fixed, changes also needed for frontend @ilmotta

File: protocol/messenger.go (hunk 5 above)
Comment (qfrank) [reply protocol/messenger.go:1739]: commits updated @ilmotta @igor-sirotin

File: protocol/messenger.go [hunk 8]
```diff
@@ -1778,88 +1785,161 @@ func (m *Messenger) handlePushNotificationClientRegistrations(c chan struct{}) {
...
+	close(filtersCh)
+	close(publicKeysCh)
+
+	select {
+	case err := <-errCh:
+		return nil, nil, err
+	default:
+	}
+
+	return m.collectResults(filtersCh, publicKeysCh)
```
Comment (igor-sirotin) [inline protocol/messenger.go]: I think `errCh` can also be passed to `collectResults` and processed inside.
Right now `collectResults` always returns `nil` error.

File: protocol/messenger.go [hunk 9]
```diff
@@ -1778,88 +1785,161 @@ func (m *Messenger) handlePushNotificationClientRegistrations(c chan struct{}) {
...
+	go m.processChats(&wg, filtersCh, publicKeysCh, errCh)
+	go m.processContacts(&wg, publicKeysCh, errCh)
+	go m.processControlledCommunities(&wg, errCh)
+
+	wg.Wait()
+	close(filtersCh)
+	close(publicKeysCh)
+
+	select {
+	case err := <-errCh:
```
Comment (igor-sirotin) [inline protocol/messenger.go]: Would be nice to log all of the errors in `errCh`.
And not just return one of the errors, but use `errors.Join` and return all of them.

File: protocol/messenger.go [hunk 10]
```diff
@@ -1778,88 +1785,161 @@ func (m *Messenger) handlePushNotificationClientRegistrations(c chan struct{}) {
...
+	return filtersToInit
+}
 
-			err = m.communitiesManager.SaveCommunitySettings(communitySettings)
-			if err != nil {
-				logger.Warn("failed to save community settings", zap.Error(err))
-			}
-			continue
-		}
+func (m *Messenger) ensureCommunitySettings(org *communities.Community) error {
```
Comment (igor-sirotin) [inline protocol/messenger.go]: Actually it looks like this function doesn't really depend on the `Messenger`, only the `CommunityManager`. Would be even better to move it there.

File: protocol/messenger.go [hunk 11]
```diff
@@ -1778,88 +1785,161 @@ func (m *Messenger) handlePushNotificationClientRegistrations(c chan struct{}) {
...
+	defer wg.Done()
+
 	spectatedCommunities, err := m.communitiesManager.Spectated()
 	if err != nil {
-		return err
+		errCh <- err
+		return
 	}
+
+	var filtersToInit []transport.FiltersToInitialize
```
Comment (igor-sirotin) [inline protocol/messenger.go]:
```suggestion
	filtersToInit := make([]transport.FiltersToInitialize, 0, len(spectatedCommunities))
```

File: protocol/messenger.go [hunk 12]
```diff
@@ -1778,88 +1785,161 @@ func (m *Messenger) handlePushNotificationClientRegistrations(c chan struct{}) {
...
 
+	filtersCh <- filters
+	publicKeysCh <- publicKeys
+
+	if err := m.processDeprecatedChats(); err != nil {
+		errCh <- err
+	}
+}
+
+func (m *Messenger) validateAndProcessChats(chats []*Chat) ([]*Chat, map[string]*communities.Community) {
```
Comment (igor-sirotin) [inline protocol/messenger.go]: This function doesn't seem to "process" chats.
```suggestion
func (m *Messenger) validateChats(chats []*Chat) ([]*Chat, map[string]*communities.Community) {
```

File: protocol/messenger.go [hunk 13]
```diff
@@ -1778,88 +1785,161 @@ func (m *Messenger) handlePushNotificationClientRegistrations(c chan struct{}) {
...
+	logger := m.logger.With(zap.String("site", "validateAndProcessChats"))
 	communityInfo := make(map[string]*communities.Community)
```
Comment (igor-sirotin) [inline protocol/messenger.go]: Not your code, but since you're around...
`communityInfo` is quite misleading, it's not just 1 community. I can see that further by the callstack it's called `communitiesCache`, this would be a better name

**Review by igor-sirotin — APPROVED** [review]
Great work splitting the function 👏 
And thanks for the separate commit with moving files, made my review much faster 👍

File: protocol/messenger_chats.go (hunk 6 above)
Comment (ilmotta) [reply protocol/messenger_chats.go]: Will apply, thanks

**Review by ilmotta — APPROVED** [review]

File: protocol/messenger.go (hunk 12 above)
Comment (qfrank) [reply protocol/messenger.go]: updated it to `validateAndInitChatsFirstMessageTimestamp`

File: services/ext/api.go [hunk 14]
```diff
@@ -291,8 +291,8 @@ func (api *PublicAPI) Chats(parent context.Context) []*protocol.Chat {
 	return api.service.messenger.Chats()
 }
 
-func (api *PublicAPI) ChatsPreview(parent context.Context) []*protocol.ChatPreview {
-	return api.service.messenger.ChatsPreview()
+func (api *PublicAPI) ChatsPreview(parent context.Context, filterType *protocol.ChatPreviewFilterType) []*protocol.ChatPreview {
```
Comment (osmaczko) [inline services/ext/api.go]: Isn't it a breaking change? :thinking:

File: protocol/messenger_filter_init.go [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+
+	filtersCh <- filters
+	publicKeysCh <- publicKeys
+
+	if err := m.processDeprecatedChats(); err != nil {
+		errCh <- err
+	}
+}
+
+func (m *Messenger) validateAndInitChatsFirstMessageTimestamp(chats []*Chat) ([]*Chat, map[string]*communities.Community) {
```
Comment (igor-sirotin) [inline protocol/messenger_filter_init.go]: oh cmon 😄 Please avoid this `and` word in the functions name. It's a sign of smth wrong. 

And in this case it's very simple to remove: 
just call `initChatsFirstMessageTimestamp` _outside_ this function, right after the call to it.

```
validChats, communitiesCache := m.valiateChats(communitiesCache, validChats)
initChatsFirstMessageTimestamp(validChats, communitiesCache)
```

Comment (qfrank) [reply protocol/messenger_filter_init.go]: 🤦‍♂️

File: protocol/messenger_chats.go [hunk 16]
```diff
@@ -46,10 +54,25 @@ func (m *Messenger) Chats() []*Chat {
...
+// When onlyCommunityChats is nil, returns all chats
+// When onlyCommunityChats is true, only returns community chats
+// When onlyCommunityChats is false, returns all non-community chats
+func (m *Messenger) ChatsPreview(filterPointer *ChatPreviewFilterType) []*ChatPreview {
```
Comment (osmaczko) [inline protocol/messenger_chats.go]: Why pointer? Isn't `ChatPreviewFilterType` covering all cases already? `nil` is redundant to `ChatPreviewFilterTypeAll`

**Review by osmaczko — APPROVED** [review]
Thanks, looks much better now.

File: services/ext/api.go (hunk 14 above)
Comment (qfrank) [reply services/ext/api.go]: use pointer, so it's not a break change for mobile, but I'm not sure if it's a break change for desktop, can you help check? @osmaczko

Comment (osmaczko) [reply services/ext/api.go]: Desktop doesn't use `ChatsPreview` endpoint AFAICS.

PS. I would prefer a breaking change for more straightforward code. Would a breaking change in this API be a significant issue? 🤔

Comment (ilmotta) [reply services/ext/api.go]: Mobile is already integrating with this updated endpoint @osmaczko https://github.com/status-im/status-mobile/pull/21642

Comment (qfrank) [reply services/ext/api.go]: okay, if desktop doesn't use it, let's break it @ilmotta

Comment (qfrank) [reply services/ext/api.go]: updated! @ilmotta

File: protocol/messenger_filter_init.go (hunk 15 above)
Comment (qfrank) [reply protocol/messenger_filter_init.go]: fixed!

File: services/ext/api.go (hunk 14 above)
Comment (osmaczko) [reply services/ext/api.go]: Thanks!

Comment (osmaczko) [reply services/ext/api.go]: Oh, what I meant is to keep the enum as it was but remove the pointer instead.

**Comment by VolodLytvynenko** [comment]
hi @qfrank could you please rebase the current PR? Thanx!

**Comment by qfrank** [comment]
> hi @qfrank could you please rebase the current PR? Thanx!

done @VolodLytvynenko
